R/validate.R

```diff
@@ -0,0 +1,120 @@
+#' Validate user-provided wastewater concentration data
+#'
+#' @param ww_data tibble containing the input wastewater data
+#' @param conc_col_name string indicating the name of the column containing
+#' the concentration measurements in the wastewater data
+#' @param lod_col_name string indicating the name of the column containing
+#' the limit of detection for each measurement in the wastewater data
+#' @param call Calling environment to be passed to [cli::cli_abort()] for
+#' traceback.
+#'
+#' @return NULL, invisibly
+validate_ww_conc_data <- function(ww_data,
+                                  conc_col_name,
+                                  lod_col_name,
+                                  call = rlang::caller_env()) {
+  ww_conc <- ww_data |> dplyr::pull({
+    conc_col_name
+  })
+  arg <- conc_col_name
+  check_non_missingness(ww_conc, arg, call)
+  check_elements_non_neg(ww_conc, arg, call)
+  checkmate::assert_vector(ww_conc)
+
+  ww_lod <- ww_data |> dplyr::pull({
+    lod_col_name
+  })
+  arg <- "lod_col_name"
+  check_non_missingness(ww_lod, arg, )
+  check_elements_non_neg(ww_lod, arg, call)
+  checkmate::assert_vector(ww_lod)
+
+  # Wastewater date column should be of date type!
+  ww_obs_dates <- ww_data$date
+  arg <- "ww_obs_dates"
+  check_non_missingness(ww_obs_dates, arg, call)
+  checkmate::assert_date(ww_obs_dates)
+  # check for multiple observations per day within a site and lab
+  ww_data |>
+    dplyr::group_by(site, lab) |>
+    check_for_repeat_elements()
+
+  # Sites  either need to be integers or characters, not be missing, and be
+  # non-negative
```

Comment:
Why do the various ids (e.g. here site ids here) ids need to be non-negative when they are ints?

We're also implicitly relying on the fact that strings can be compared to numbers for the non-negativity check to pass. That seems potentially a bit opaque, and could have unexpected/surprising behavior. Consider for example:
```R
> "a" >= 0
[1] TRUE
> "-a" >= 0
[1] FALSE
> "a" >= Inf
[1] FALSE
> "z" >= Inf
[1] TRUE
```

Explanation: https://stackoverflow.com/a/57090193

Reply:
Hmm I hadn't realized that behavior, they probably don't need to be checked for non-negativity, just integer or character!

Reply:
Related question: presumably any character-based indices do need to get converted to integer-based ones somewhere before they are passed to Stan. Where in the codebase does that happen, and do we have validation and/or unit testing for it?

Reply:
This happens in the `get_stan_data()` function -- we probably need to add unit testing to check for the ability to do this again there!